File: 2024-10-October-GFG-POTD/October 13 - Delete Alternate Nodes/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int x) {
        data = x;
        next = nullptr;
    }
};

class Solution {
public:
    void deleteAlt(Node* head) {
        if (head == nullptr) return;  // if the list is empty, nothing to do

        Node* current = head;

        // Traverse the list and delete every alternate node
        while (current != nullptr && current->next != nullptr) {
            // Skip the next node by linking current node to the one after the next
            current->next = current->next->next;
            // Move to the next non-deleted node
            current = current->next;
        }
    }
};

// Function to print the linked list
void printList(Node* node) {
    while (node != nullptr) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << endl;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t-- > 0) {
        vector<int> arr;
        string input;
        getline(cin, input);
        istringstream ss(input);
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }

        Node* head = new Node(arr[0]);
        Node* tail = head;
        for (int i = 1; i < arr.size(); i++) {
            tail->next = new Node(arr[i]);
            tail = tail->next;
        }

        Solution ob;
        ob.deleteAlt(head);
        printList(head);
    }

    return 0;
}

File: 2024-10-October-GFG-POTD/October 13 - Delete Alternate Nodes/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def deleteAlt(self, head):
        if head is None:
            return  # If the list is empty, nothing to do

        current = head

        # Traverse the list and delete every alternate node
        while current is not None and current.next is not None:
            # Skip the next node by linking current node to the one after the next
            current.next = current.next.next
            # Move to the next non-deleted node
            current = current.next

# Function to print the linked list
def printList(node):
    while node is not None:
        print(node.data, end=" ")
        node = node.next
    print()

if __name__ == "__main__":
    t = int(input())
    
    for _ in range(t):
        arr = list(map(int, input().split()))
        
        head = Node(arr[0])
        tail = head
        for i in range(1, len(arr)):
            tail.next = Node(arr[i])
            tail = tail.next

        ob = Solution()
        ob.deleteAlt(head)
        printList(head)
